File: 프로그래머스/3/150366. 표 병합/표 병합.cpp
#include <bits/stdc++.h>
using namespace std;


struct cell{
    int parent;
    string value = "EMPTY";
};

vector<cell> table;
int Find(int num){

    if(table[num].parent == num){
        return num;
    }

    return table[num].parent = Find(table[num].parent);
}

void Union(int a, int b){
    a = Find(a); b = Find(b);

    if(table[a].value == "EMPTY" && table[b].value != "EMPTY"){
        table[a].parent = b;
    }else{
        table[b].parent = a;
    }
}

vector<string> solution(vector<string> commands) {
    vector<string> answer;

    table.resize(50*50);

    for(int i = 0; i<table.size(); i++) table[i].parent = i;

    for(string s: commands){
        vector<string> tmp;
        string token;
        istringstream iss(s);
        while(iss >> token){
            tmp.push_back(token);
        }

        if(tmp[0] == "UPDATE"){
            if(tmp.size() == 4){
                int r = stoi(tmp[1]) - 1;
                int c = stoi(tmp[2]) - 1;
                int idx = r * 50 + c;
                table[idx].value = tmp[3];
            }else{
                for(auto i : table){
                    if(i.value == tmp[2]) i.value = tmp[3];
                }
            }
        }else if(tmp[0] == "MERGE"){
            int r1 = stoi(tmp[1]) - 1;
            int c1 = stoi(tmp[2]) - 1;
            int r2 = stoi(tmp[3]) - 1;
            int c2 = stoi(tmp[4]) - 1;

            int idx1 = r1*50+c1;
            int idx2 = r2*50+c2;

            Union(idx1, idx2);
        }else if(tmp[0] == "UNMERGE"){
            int r = stoi(tmp[1]) - 1;
            int c = stoi(tmp[2]) - 1;
            int idx = r * 50 + c;

            string val = table[idx].value;
            int root = Find(idx);

            vector<int> target;
            for(int i = 0; i<table.size(); i++){
                if(Find(i) == root) target.push_back(i);
            }
            for(auto i : target){
                table[i].parent = i;
                table[i].value = "EMPTY";
            }
            table[idx].parent = idx;
            table[idx].value = val;
        }else{
            int r = stoi(tmp[1]) - 1;
            int c = stoi(tmp[2]) - 1;
            int idx = r*50 + c;
            answer.push_back(table[idx].value);
        }
    }

    return answer;
}

File: 프로그래머스/3/150366. 표 병합/표 병합.java
import java.util.*;


class UF {
    List<List<int[]>> parent;
    String[][] val;
    
    public UF(){
        parent = new ArrayList<>();
        for(int i = 0; i<51; i++){
            parent.add(new ArrayList<>());
        }
        
        for(int i = 0; i<51; i++){
            for(int j = 0; j<51; j++){
                parent.get(i).add(new int[]{i,j});
            }
        }
        
        val = new String[51][51];
    }
    
    
    void merge(int x1, int y1, int x2, int y2){
        int[] r1 = find(x1, y1);
        int[] r2 = find(x2, y2);
        
        if(r1[0] == r2[0] && r1[1] == r2[1]) return;
        
        String v1 = val[r1[0]][r1[1]];
        String v2 = val[r2[0]][r2[1]];
        String keep = (v1 != null) ? v1 : v2;
        
        parent.get(r2[0]).set(r2[1], r1);
        val[r1[0]][r1[1]] = keep;
        val[r2[0]][r2[1]] = null;
    }
    
    void unmerge(int x, int y){
        
        int[] root = find(x, y);
        String value = val[root[0]][root[1]];
        
        List<int[]> members = new ArrayList<>();
        for (int i = 1; i <= 50; i++) {
            for (int j = 1; j <= 50; j++) {
                int[] r = find(i, j);
                if (r[0] == root[0] && r[1] == root[1]) {
                    members.add(new int[]{i, j});
                }
            }
        }
        
        for(int[] cell : members) {
            int i = cell[0], j = cell[1];
            parent.get(i).set(j, new int[]{i,j});
            val[i][j] = null;
        }
        
        val[x][y] = value;
    }
    
    void update(int x, int y, String value){
        int[] root = find(x, y);
        
        val[root[0]][root[1]] = value;
    }
    
    void updateFromVal(String val1, String val2){
        for (int i = 1; i <= 50; i++) {
            for (int j = 1; j <= 50; j++) {
                int[] r = find(i, j);
                String cur = val[r[0]][r[1]];
                if (val1.equals(cur)) {
                    val[r[0]][r[1]] = val2;
                }
            }
        }
    }
    
    int[] find(int x, int y) {
        int[] cur = parent.get(x).get(y);
        if (cur[0] == x && cur[1] == y) return cur;
        int[] root = find(cur[0], cur[1]);
        parent.get(x).set(y, root);
        return root;
    }
    
    String print(int x, int y){
        int[] root = find(x, y);
        
        return findValue(root[0], root[1]);
    }
    
    String findValue(int x, int y){
        return val[x][y];
    }
}


class Solution {
    public String[] solution(String[] commands) {
        String[] answer;
        
        UF uf = new UF();
        List<String> ans = new ArrayList<>();
        for(String command: commands){
            String[] tmp = command.split(" ");
            String c = tmp[0];
            
            if(c.equals("UPDATE") && tmp.length == 4) { 
                int x = Integer.parseInt(tmp[1]);
                int y = Integer.parseInt(tmp[2]);
                String val = tmp[3];
                
                uf.update(x, y, val);
            }else if(c.equals("MERGE")){
                int x1 = Integer.parseInt(tmp[1]);
                int y1 = Integer.parseInt(tmp[2]);
                int x2 = Integer.parseInt(tmp[3]);
                int y2 = Integer.parseInt(tmp[4]);
                
                uf.merge(x1, y1, x2, y2);
                
            }else if(c.equals("UPDATE")){
                String val1 = tmp[1];
                String val2 = tmp[2];
                
                uf.updateFromVal(val1, val2);
            }else if(c.equals("UNMERGE")){
                int x1 = Integer.parseInt(tmp[1]);
                int y1 = Integer.parseInt(tmp[2]);
                uf.unmerge(x1, y1);
            } else {
                int x1 = Integer.parseInt(tmp[1]);
                int y1 = Integer.parseInt(tmp[2]);
                String res = uf.print(x1, y1);
                res = (res == null || res.equals(""))? "EMPTY" : res;
                ans.add(res);
            }   
        }
        
        answer = new String[ans.size()];
        for(int i = 0; i<ans.size(); i++){
            answer[i] = ans.get(i);
        }
        return answer;
    }
}